### PR TITLE
[WIP] Figure out the most ergonomic way to consume libp2p

### PR DESCRIPTION
**Pull request by dryajov**
This is a draft PR to come up with the most ergonomic way of consuming nim-libp2p.

### CODE REVIEW

**Comment by stefantalpalaru** [comment]
Pinging @tina1998612 who will need to change an article, after this is merged.

File: tests/testswitch.nim [hunk 1]
```diff
@@ -2,23 +2,7 @@ import unittest, tables
...
-                  crypto/crypto,
-                  peer,
-                  protocols/protocol,
-                  muxers/muxer,
-                  muxers/mplex/mplex,
-                  muxers/mplex/types,
-                  protocols/secure/secio,
-                  protocols/secure/secure,
-                  stream/lpstream]
+import ../libp2p/libp2p
```
Comment (stefantalpalaru) [inline tests/testswitch.nim:5]: Or `../libp2p/all` to avoid the stutter.

File: libp2p.nim [hunk 2]
```diff
@@ -8,10 +8,11 @@
...
 
 import
+  libp2p/libp2p,
   libp2p/daemon/[daemonapi, transpool],
   libp2p/protobuf/minprotobuf,
   libp2p/varint
 
 export
-  daemonapi, transpool, minprotobuf, varint
+  libp2p, daemonapi, transpool, minprotobuf, varint
```
Comment (dryajov) [inline libp2p.nim:17]: weird that I can't do 

```nim
import libp2p

let switch = newSwitch(...)
```

does seem to reexport properly

File: tests/testswitch.nim (hunk 1 above)
Comment (dryajov) [reply tests/testswitch.nim:5]: I'm trying to export it through the main `libp2p` file in the root of the project, but doesn't seem to work - https://github.com/status-im/nim-libp2p/pull/133#pullrequestreview-389542329

**Comment by tina1998612** [comment]
No problem~ thanks for reminding

**Comment by cheatfate** [comment]
The main idea is to create not only simple `import libp2p` but also keep plugin architecture.
So the core functionality should be available with `libp2p/core`.
This module should import and re-export
```
switch.nim
connection.nim
multistream.nim
varint.nim
vbuffer.nim
wire.nim
transcoder.nim
utility.nim
peer.nim
peerinfo.nim
multihash.nim
multiaddress.nim
multibase.nim
multicodec.nim
cid.nim
stream/bufferstream.nim
stream/chronosstream.nim
stream/lpstream.nim
protobuf/minprotobuf.nim
crypto/*
```
All other things should be pluggable and `switch` should react on protocols imported via `when defined()` things.

**Comment by dryajov** [comment]
> The main idea is to create not only simple import libp2p but also keep plugin architecture.

@cheatfate I agree and at first I played around with that approach, however I quickly realized that "official" module (muxers, pubsub, discovery, etc...) would be part of this repo either way, so excluding them doesn't make much sense and complicates the experience, because you need to know where in the directory tree this components are located. Perhaps there is a penalty in importing everything and that should definitely be considered, but it seems a lot more convenient this way. 

In any case, this is more of a draft PR right now, so all input is appreciated :)

**Comment by stefantalpalaru** [comment]
Make it run standalone, with a nim-libp2p installed using Nimble. This way, people can just copy/paste import lines from it.

For the CI, you can just specify an include path on the `nim` command line.

**Comment by dryajov** [comment]
I've been thinking about this some more and the issue isn't only about sane exports/imports, it also touches on node construction, which right now is pretty cumbersome as well. I think that maybe something like the following could be an option.

```nim
import libp2p # single import, re-exports everything

proc createMplex(conn: Connection) =
  # custom instance provider
  newMplex(conn)

var peerInfo = PeerInfo.init(...)

var node = libp2p.with(Secio)
                 .with(Mplex, createMplex) # second param will be used as instance provider
                 .with(GossipSub)
                 .with(TcpTransport)
                 .init(peerInfo)
```

**Comment by dryajov** [comment]
> This PR is not taking in to account docs/tutorial at all, they use still a wrong relative path.

We can update those, the API is still alpha and I'm sure will continue evolving some more, so nothing to grief about IMO...

**Comment by sinkingsugar** [comment]
> > This PR is not taking in to account docs/tutorial at all, they use still a wrong relative path.
> 
> We can update those, the API is still alpha and I'm sure it will continue to evolve some more, so nothing to grief about IMO...

I get it but the point is, do we care about nimble or not ? mostly I guess, or assume people pull / clone us?

**Comment by stefantalpalaru** [comment]
You can safely assume that most people will install this library using Nimble.

**Comment by dryajov** [comment]
> You can safely assume that most people will install this library using Nimble.

I'm not sure how this will preclude it from happening.

**Comment by stefantalpalaru** [comment]
You shouldn't have relative imports in an example from which people might copy/paste. What will happen if they copy `import ../libp2p/libp2p` from an article or from GitHub and try to use it with a libp2p instance installed with Nimble?

**Comment by dryajov** [comment]
apparently you can have a `config.nims` with this line:

```nim
switch("path", "$projectDir/../src")
```

that will adjust relative paths and it would be possible to do imports without relative paths.

**Comment by tina1998612** [comment]
Will this be merged soon? So that I can update the articles accordingly ~ thanks!

**Comment by arnetheduck** [comment]
so, regarding a lot of these imports and exports, the recommendation would be that whenever something public uses a type from a different module, that module should be reexported - for example, if a `HashSet` is returned from a public function, `std/sets` should be exported.

once this is done consistently, the list of imports will generally be much shorter, and then one can start looking at whether more needs to be done

**Comment by arnetheduck** [comment]
it is also generally better to use explicit folders in imports - ie `./some_module` as this makes  sure that the right module gets imported even if  a module with the same name appears in some random dependency -

**Comment by arnetheduck** [comment]
shall we close this as  obsolete?

**Pull request closed without merging**